File: jScrap/src/main/java/com/sorception/jscrap/services/TokenService.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package com.sorception.jscrap.services;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.sorception.jscrap.dao.IGenericDAO;
import com.sorception.jscrap.dao.ITokenDAO;
import com.sorception.jscrap.entities.TokenEntity;
import com.sorception.jscrap.error.ResourceNotFoundException;
import com.sorception.jscrap.webservices.SGClient;

/**
 *
 * @author kaseyo
 */
@Service
@Transactional(noRollbackFor={ResourceNotFoundException.class})
public class TokenService extends AbstractService<TokenEntity> {
	@Autowired
    private SGClient sgClient;
    
    @Autowired
    private ActiveMQService activeMQService;
    
    @Autowired
    private ITokenDAO dao;
    
    public TokenService() {
		super(TokenEntity.class);
	}
    
	@Override
	protected IGenericDAO<TokenEntity> getDao() {
		return dao;
	}
	
    public TokenEntity requestToken() {
        logger.info("New token requested. Disabling current token and "
        		+ "AMQ services and requesting new one from web service.");
        getTokenDao().invalidateTokens();
        // Access to web service
        TokenEntity temporalToken = sgClient.signUp();
        // Disable jmsContainer
        activeMQService.disableJmsContainers();
        // Save temporal token
        return create(temporalToken);
    }
    
    public TokenEntity getValid() {
    	logger.info("Retrieving current token...");
        TokenEntity tokenEntity = getTokenDao().findByStatus(TokenEntity.TokenStatus.VALID);
        if(null == tokenEntity) {
            // Check if we have requested one
            tokenEntity = getRequestOrTemporal();
            if(null == tokenEntity) { // If not, throw 404
            	logger.info("Not valid token has been found.");
                throw new ResourceNotFoundException("Not valid token or request found.");
            }
            // Check if new token is available
            // Method getState will throw NotFound if not valid
            TokenEntity newToken = sgClient.getState(tokenEntity.getToken());
            tokenEntity = create(newToken);
            if(!tokenEntity.isValid()) { // If not, throw 404
            	logger.info("Token request has not been accepted.");
                throw new ResourceNotFoundException("Token request has not been accepted.");
            } else // Enable JmsContainer
            	activeMQService.enableJmsContainers(tokenEntity);
        }
        return tokenEntity;
    }
    
    public List<TokenEntity> list() {
    	logger.info("Retrieving all tokens.");
        return getTokenDao().findAllOrderByCreated();
    }
    
    protected ITokenDAO getTokenDao() {
    	return ((ITokenDAO)getDao());
    }
    
    protected TokenEntity getRequestOrTemporal() {
    	List<TokenEntity> tokens = getTokenDao()
    			.findByStatus(TokenEntity.TokenStatus.REQUESTED, TokenEntity.TokenStatus.TEMPORAL);
    	if(tokens.size() > 0)
    		return tokens.get(0);
    	else
    		return null;
    }
}

File: Scrap.NET/ScrapWeb/Services/TokenService.cs
using ScrapWeb.DataAccess;
using ScrapWeb.Entities;
using ScrapWeb.Exceptions;
using ScrapWeb.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Web;

namespace ScrapWeb.Services
{
    public class TokenService
    {
        private ScrapContext scrapContext;
        private GenericRepository<TokenEntity> tokenRepository;

        private SGService sgService;
        private AMQService amqService;

        public TokenService()
        {
            scrapContext = new ScrapContext();
            tokenRepository = new GenericRepository<TokenEntity>(scrapContext);
            sgService = new SGService();
            amqService = new AMQService();
        }

        public IEnumerable<TokenEntity> getAll() {
            return tokenRepository.GetAll();
        }

        public TokenEntity requestToken()
        {
            // Set valid tokens to null
            nullValidTokens();

            // Disable topic subscriber
            amqService.destroyTopicSubscribers();

            // Request new token as temporal
            TokenEntity tokenEntity = sgService.signUp();
            tokenRepository.Insert(tokenEntity);

            // Save changes and return
            scrapContext.SaveChanges();
            return tokenEntity;
        }

        public TokenEntity getValid()
        {
            var tokenEntity = tokenRepository
                .Get(t => t.status == TokenStatus.VALID)
                .FirstOrDefault();
            if (tokenEntity == null)
            {
                // Check if we have requested one
                tokenEntity = tokenRepository
                    .Get(t => (t.status == TokenStatus.REQUESTED || t.status == TokenStatus.TEMPORAL))
                    .FirstOrDefault();
                if (tokenEntity == null)
                    throw new ServiceException("Not valid token or request was found", HttpStatusCode.NotFound);
                // Check if token is available
                tokenEntity = sgService.getState(tokenEntity.token);
                tokenRepository.Insert(tokenEntity);
                // Save changes
                scrapContext.SaveChanges();
                // If not available, 404
                if (tokenEntity.status != TokenStatus.VALID)
                    throw new ServiceException("Token request has not been accepted", HttpStatusCode.NotFound);
                // If we have a token, enable it here
                amqService.createTopicSubscribers(tokenEntity);
            }
            return tokenEntity;
        }

        private void nullValidTokens()
        {
            var tokens = tokenRepository.Get(t => t.status == TokenStatus.VALID);
            foreach (var token in tokens)
            {
                token.status = TokenStatus.EXPIRED;
                tokenRepository.Update(token);
            }
        }
    }
}